### PR TITLE
Enable saved views feature in app

### PR DESCRIPTION
**Pull request by kaixi-wang**
### Add new feature: saved views
- this branch is rebased on top of https://github.com/voxel51/fiftyone/tree/saved-views, which added SDK support 

#### Feature checklist
- [x] Saving a view works as expected
    - [x] OSS
        - [x] Saves the new view [done]
        - [x] Changes the url to new name (if name changed) [done]
    - [x] Teams
        - [x] Saves the new view [done]
        - [x] Changes the url to new name (if name changed) [done]

- [x] Selecting a view works as expected
    - [x] OSS
        - [x] Selects the new view content - changes the view [done]
        - [x] Changes the URL [done]
    - [x] OSS Multi-tab
        - [x] Selects the new view content - and communicates to other tab [done]
        - [x] Clearing the view bar clears the view and resets saved view selection to None [done]
        - [x] Touching an already saved view resets url (removes view query param if any) and resets saved view selection to None [done]
        - [x] Creating a new view, deleting an existing view, and updating existing views syncs between tabs automatically. [WON"T DO - refresh in other tabs is sufficient]
    - [x] Teams
        - [x] Changes URL [done]
        - [x] Changes view [done]

- [x] Clearing the view
    - [x] OSS
        - [x] Clears the view and loads all samples [done]
        - [x] Removes the url param view (if any) and deselects the selected view in saved view selection [done]
    - [x] Teams
        - [x] Clears the view and loads all samples [done]
        - [x] Removes the url param view (if any) and deselects the selected view in saved view selection [done]

- [x] Extending a view
    - [x] OSS
        - [x] should remove view query param from url, deselect saved view, BUT keep view bar functional as before. also update content/samples [done]
    - [x] Teams
        - [x] should remove view query param from url, deselect saved view, BUT keep view bar functional as before. also update content/samples [done]

- [x] Updating a view
    - [x] OSS
        - [x] Saves the new updated view [done]
        - [x] Changes url if name changed [done]
        - [x] View selection gets updated [done]
    - [x] Teams
        - [x] Saves the new updated view [done]
        - [x] Changes url if name changed [done]
        - [x] View selection gets updated [done]

- [ ] Deleting a view
    - [x] OSS
        - [x] Saves the new updated view [done]
        - [x] Changes url if name changed [done]
        - [x] View selection gets updated [done]
    - [ ] Teams
        - [x] Removes the saved view [done]
        - [x] Changes url if name changed [done]
        - [ ] View selection gets updated [does NOT - needs investigation]

- [x] Directly linking to an existing view should load the view successfully
    - [x] OSS [done]
    - [x] Teams [done]

- [x] Directly linking to a non-existing view (?view=non-existing-view-slug) should clear the view content and clear the saved view selection
    - [x] OSS [done]
    - [x] Teams [done]

### Additional 

- [x] Resolve `DictField` and `ListField` type reference fields when serializing `Document` via `to_dict()` to support GraphQL queries for saved views, brain methods, annotation runs, and evaluations

## How is this patch tested? If it is not, please explain why.

(Details)

### Is this a user-facing change that should be mentioned in the release notes?

-   [ ] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

### CODE REVIEW

File: fiftyone/server/mutation.py [hunk 1]
```diff
@@ -135,3 +135,20 @@ async def set_group_slice(
...
+        description: t.Optional[str] = None,
+        color: t.Optional[str] = None,
+    ) -> bool:
+        state = get_state()
+        state.dataset.save_view(
+            view_name, state.view, description=description, color=color
+        )
+        state.view = state.dataset.load_view(view_name)
+        await dispatch_event(subscription, StateUpdate(state=state))
+        return True
```
Comment (manivoxel51) [inline fiftyone/server/mutation.py:154]: we should return the newly saved view.

**Review by manivoxel51 — COMMENTED** [review]
nice! left a small comment

File: app/packages/core/src/Root/Root.tsx [hunk 2]
```diff
@@ -53,6 +53,29 @@ const rootQuery = graphql`
...
 
+export const DatasetSavedViewsQuery = graphql`
+  query RootDatasetSavedViewsQuery($name: String!) {
+    ...RootDatasetSavedViewsFragment
+  }
+`;
+
+export const DatasetSavedViewsFragment = graphql`
+  fragment RootDatasetSavedViewsFragment on Query
+  @refetchable(queryName: "EootDatasetSavedViewsQuery") {
```
Comment (kaixi-wang) [inline app/packages/core/src/Root/Root.tsx:64]: @manivoxel51 typo?
Can this be changed to `RootDatasetSavedViewsQuery` and get rid of `export const DatasetSavedViewsQuery...`? You could then use RootDatasetSavedViewsQuery$key or something along those lines in the places where you call `useRefetchableFragment`

File: fiftyone/core/runs.py [hunk 3]
```diff
@@ -365,14 +365,18 @@ def save_run_info(cls, samples, run_info, overwrite=True):
...
+        run_doc = RunDocument(
+            dataset_id=dataset_doc.id,
             key=key,
             version=run_info.version,
             timestamp=run_info.timestamp,
             config=deepcopy(run_info.config.serialize()),
             view_stages=view_stages,
             results=None,
         )
+        # run_doc.save()
```
Comment (kaixi-wang) [inline fiftyone/core/runs.py:377]: @brimoor should the run_doc be saved? This was originally in the saved_views pr but not in the develop head now

File: app/packages/core/src/Root/Root.tsx (hunk 2 above)
Comment (manivoxel51) [reply app/packages/core/src/Root/Root.tsx]: agreed we should fix the typo - fixed!
agreed that we could consolidate the query - can tackle that later though

File: .prettierrc.js [hunk 4]
```diff
@@ -17,7 +17,7 @@ module.exports = {
     {
       files: "*.json",
       options: {
-        tabWidth: 4,
+        tabWidth: 2,
```
Comment (manivoxel51) [inline .prettierrc.js:20]: nit: is this part of our saved-views change? 🤔

File: app/packages/app/src/index.tsx [hunk 5]
```diff
@@ -137,7 +169,7 @@ const App: React.FC = ({}) => {
     );
 
     return () => controller.abort();
-  }, []);
+  }, [clearModal, handleError, refresh, screenshot, subscription]);
```
Comment (manivoxel51) [inline app/packages/app/src/index.tsx]: @kaixi-wang I am removing these dependencies - they seem to cause the samples modal **not** to open.

File: app/packages/core/src/components/ViewBar/ViewBar.tsx [hunk 6]
```diff
@@ -83,21 +83,24 @@ const ViewBar = React.memo(() => {
       setView,
       fieldNames: fieldPaths,
     });
-  }, [view]);
+  }, [view, fieldPaths, send]);
```
Comment (manivoxel51) [inline app/packages/core/src/components/ViewBar/ViewBar.tsx:86]: @kaixi-wang did we add these? I'd say let's hold off on adding dependencies to useEffects unless they fix a specific issue. useEffect has side effects and in general I like to test the entire app when dependencies get added to them. I remove them. lmk if you object

File: app/packages/core/src/components/ViewBar/ViewStage/ViewStageParameter.tsx [hunk 7]
```diff
@@ -189,24 +189,27 @@ const ObjectEditor = ({
 
   useEffect(() => {
     containerSet({
-      position: state.matches("editing") ? "fixed" : "relative",
-      backgroundColor: state.matches("editing")
+      position: state?.matches("editing") ? "fixed" : "relative",
```
Comment (manivoxel51) [inline app/packages/core/src/components/ViewBar/ViewStage/ViewStageParameter.tsx:192]: hm... did we change this file? 🤔

File: app/packages/state/src/routing/RouterContext.ts [hunk 8]
```diff
@@ -255,7 +255,13 @@ async function fetchGraphQL(
   );
 
   if ("errors" in data && data.errors) {
-    throw new GraphQLError(data.errors as unknown as GraphQLError[]);
+    // TODO: figure out how why the aggregationQuery is getting
```
Comment (manivoxel51) [inline app/packages/state/src/routing/RouterContext.ts:258]: 👍 nice!

File: app/schema.graphql [hunk 9]
```diff
@@ -406,6 +444,27 @@ type SampleItemStrPageInfo {
...
+  name: String
+  slug: String
+  description: String
+  color: String
+  viewStages: [String!]
+  createdAt: datetime
+  lastModifiedAt: datetime
+  lastLoadedAt: datetime
+  viewName: String
+  stageDicts: BSONArray
```
Comment (manivoxel51) [inline app/schema.graphql:459]: do we need this?

File: .prettierrc.js (hunk 4 above)
Comment (manivoxel51) [reply .prettierrc.js:20]: regardless totally cool

File: app/packages/state/src/recoil/aggregations.ts [hunk 10]
```diff
@@ -76,9 +79,19 @@ export const aggregation = selectorFamily({
...
     }) =>
     ({ get }) => {
-      return get(
+      const result = get(
         aggregations({ ...params, paths: get(schemaAtoms.filterFields(path)) })
-      ).filter((data) => data.path === path)[0];
+      ).filter((data) => data.path === path);
+      // Avoid downstream errors due to undefined.map by returning an
+      // object for failed graphQL aggregations
+      return result
```
Comment (manivoxel51) [inline app/packages/state/src/recoil/aggregations.ts:87]: this should be `result.length` right?

File: .prettierrc.js (hunk 4 above)
Comment (kaixi-wang) [reply .prettierrc.js:20]: not part of saved views, but I'm tired of always having the file reformat back and forth and making diffs longer than they actually are so I changed it

File: app/packages/core/src/components/ViewBar/ViewBar.tsx (hunk 6 above)
Comment (kaixi-wang) [reply app/packages/core/src/components/ViewBar/ViewBar.tsx]: the react hooks linter said to add them so I did and it seemed to make things more stable on my side... Also reading up on hooks, it sounded like we need to be adding them https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies...

File: app/packages/state/src/recoil/aggregations.ts (hunk 10 above)
Comment (kaixi-wang) [reply app/packages/state/src/recoil/aggregations.ts]: yeah good catch

File: app/packages/core/src/components/ViewBar/ViewBar.tsx (hunk 6 above)
Comment (manivoxel51) [reply app/packages/core/src/components/ViewBar/ViewBar.tsx]: I agree with you and linter :) to keep the risks low (like the sample modal not opening in another thread in the pr) I simply suggested to keep dependency changes minimal unless they fix a specific scenario with saved-views

**Pull request closed without merging**